Issue#134 add contact entity to user

Abhai2016 opened the pull request:
dev
## ZenHub

* [Main ZenHub ticket](https://app.zenhub.com/workspaces/home-project-5f7b77ff8db73a001cb17009)


## Code reviewers

- [ ] @github_username

### Second Level Review

- [ ] @github_username

## Summary of issue

ToDo

## Summary of change

ToDo

## Testing approach

ToDo

## CHECK LIST
- [ ]  СI passed
- [ ]  Сode coverage >=95%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

Commit:
Issue#134: Added contacts to user

Commit:
Issue#134: Fixed checkstyle for data-migration

Commit:
Issue#134: Deleted contact_type from entity and dto

Commit:
Issue#134: Changed version of rsql

Commit:
Issue#134: Added contacts to rsql

Commit:
Issue#134: Deleted redundant configs and annotations

Commit:
Issue#134: Got rid of lombok.config by chanings mapper configs

Commit:
Issue#134: Deleted redundant annotations from mapper configs

Commit:
Issue#134: Added check for cotnact_type in payload and database

Commit:
Issue#134: Fixed checkstyle mistakes

Commit:
Issue#134: Simplified mapping configs

Commit:
Merge branch 'dev' into Issue#134-Add-Contact-Entity-To-User

# Conflicts:
#	home-open-api/src/main/resources/yaml/openapi.yaml

Commit:
Issue#134: Added query for contactd

Commit:
Issue#134: Changed primary on main and contactEmail on email(same for phone)

Commit:
Issue#134: Got back old realization of mapping configs

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
-        - name: "id"
+        - name: "users_id"
+          in: "path"
+          description: "Use 1 for testing."
```
Please avoid such a description. Let's provide the following one:
`The ID of the User.`

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+    post:
+      tags:
+        - contact
+      summary: "Create new contacts"
```
`Create a new contact for the User.`

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+      tags:
+        - contact
+      summary: "Create new contacts"
+      description: "Create new contacts"
```
`Create a new contact for the specified user ID.`

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+        - contact
+      summary: "Create new contacts"
+      description: "Create new contacts"
+      operationId: createContacts
```
`createContactOnUser`

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+      parameters:
+        - name: "users_id"
+          in: "path"
+          description: "Use 1 for testing."
```
`The ID of the User.`

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
       operationId: removeUser
       parameters:
-        - name: "id"
+        - name: "users_id"
```
`user_id`

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+      description: "Create new contacts"
+      operationId: createContacts
+      parameters:
+        - name: "users_id"
```
`user_id`

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+            format: int64
+            example: 1
+      requestBody:
+        description: Create a new Contact
```
`A payload for creating a contact.`

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+        required: true
+      responses:
+        '201':
+          description: Successful created
```
`Successfully created`

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+    get:
+      tags:
+        - contact
+      summary: "Get all contacts by user"
```
`Get contacts for the user`

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+      tags:
+        - contact
+      summary: "Get all contacts by user"
+      description: "Get all contacts by user"
```
`Retrieves the list of contacts associated with the specified user ID.`

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+        - contact
+      summary: "Get all contacts by user"
+      description: "Get all contacts by user"
+      operationId: getAllContacts
```
`queryContactsOnUser`

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+      description: "Get all contacts by user"
+      operationId: getAllContacts
+      parameters:
+        - name: "users_id"
```
`user_id`

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+      parameters:
+        - name: "users_id"
+          in: "path"
+          description: "Use 1 for testing."
```
The ID of the User.

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+            default: 5
+          required: false
+          description: The number of items to return.
+        - in: query
```
Why did you omit the description for all of the query parameters?

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+            type: string
+      responses:
+        '200':
+          description: successful operation
```
`Success`

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+      tags:
+        - contact
+      summary: "Get contact by id"
+      description: "Get contact by id"
```
Retrieves an existing contact associated with the specified user by user and contact IDs.

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+        - contact
+      summary: "Get contact by id"
+      description: "Get contact by id"
+      operationId: getContact
```
getContactOnUser

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+      description: "Get contact by id"
+      operationId: getContact
+      parameters:
+        - name: "users_id"
```
`user_id`

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+      parameters:
+        - name: "users_id"
+          in: "path"
+          description: "Use 1 for testing."
```
...

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+            type: integer
+            format: int64
+            example: 1
+        - name: "contacts_id"
```
`contact_id`

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+            example: 1
+        - name: "contacts_id"
+          in: "path"
+          description: "Use 1 for testing."
```
...

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+            example: 1
+      responses:
+        '200':
+          description: successful operation
```
...

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+        - contact
+      summary: "Update current contact"
+      description: "Update current contact"
+      operationId: updateContact
```
updateContactOnUser

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+    put:
+      tags:
+        - contact
+      summary: "Update current contact"
```
Update contact

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+      tags:
+        - contact
+      summary: "Update current contact"
+      description: "Update current contact"
```
Updates information on a contact that exists in the system by the specified user and contact IDs.

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+      description: "Update current contact"
+      operationId: updateContact
+      parameters:
+        - name: "users_id"
```
...

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+      parameters:
+        - name: "users_id"
+          in: "path"
+          description: "Use 1 for testing."
```
...

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+            type: integer
+            format: int64
+            example: 1
+        - name: "contacts_id"
```
...

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+            example: 1
+        - name: "contacts_id"
+          in: "path"
+          description: "Use 1 for testing."
```
...

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+            format: int64
+            example: 1
+      requestBody:
+        description: Update current contact
```
The payload for updating an existing contact.

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+        required: true
+      responses:
+        '200':
+          description: successful operation
```
Success

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+          $ref: '#/components/responses/Forbidden'
+        "404":
+          $ref: '#/components/responses/NotFound'
+        "409":
```
When we can get this status?

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+    delete:
+      tags:
+        - contact
+      summary: "Delete contact by ID"
```
`Remove contact`

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+      tags:
+        - contact
+      summary: "Delete contact by ID"
+      description: "Delete contact by ID"
```
Removes an existing contact on a user associated with the specified user and contact IDs.

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+        - contact
+      summary: "Delete contact by ID"
+      description: "Delete contact by ID"
+      operationId: removeContact
```
removeContactOnUser

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+      description: "Delete contact by ID"
+      operationId: removeContact
+      parameters:
+        - name: "users_id"
```
...

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+      parameters:
+        - name: "users_id"
+          in: "path"
+          description: "Use 1 for testing."
```
...

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+            type: integer
+            format: int64
+            example: 1
+        - name: "contacts_id"
```
...

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+            example: 1
+        - name: "contacts_id"
           in: "path"
           description: "Use 1 for testing."
```
...

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -382,7 +388,243 @@ paths:
+      summary: "Get all contacts by user"
+      description: "Get all contacts by user"
+      operationId: getAllContacts
+      parameters:
```
Please add `type` query parameter. So we can search contact by its type.

kh0ma commented on home-open-api/src/main/resources/yaml/openapi.yaml:
```diff
@@ -607,11 +846,106 @@ components:
+      properties:
+        phone:
+          type: string
+          example: "+380966617305"
```
Please add newline at the end of the file

kh0ma commented on home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/ContactServiceImpl.java:
```diff
@@ -0,0 +1,104 @@
+        return mapper.convert(contact, ContactDto.class);
+    }
+
+    @Transactional
```
Do we really need a transaction here?

Abhai2016 replied:
We needed it before because it didn't work without it but now it works fine without this annotation so I deleted it. Thanks for paying attendtion on it.

kh0ma commented on home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/ContactServiceImpl.java:
```diff
@@ -0,0 +1,104 @@
+        if (contactOptional.isPresent()) {
+            Contact contact = contactOptional.get();
+            ContactDto contactDto = mapper.convert(contact, ContactDto.class);
+            if (contactDto.getContactType().equals(updateContactDto.getContactType())) {
```
Can we change?
```
            ContactDto contactDto = mapper.convert(contact, ContactDto.class);
            if (contactDto.getContactType().equals(updateContactDto.getContactType())) {
```
to
```
            ContactTypeDto existingContactType = mapper.convert(contact.getType(), ContactTypeDto.class);
            if (existingContactType == updateContactDto.getContactType()) {
```

Abhai2016 replied:
Done.

kh0ma commented on home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/ContactServiceImpl.java:
```diff
@@ -0,0 +1,104 @@
+            if (contactDto.getContactType().equals(updateContactDto.getContactType())) {
+                return checkContactType(contact, updateContactDto);
+            } else {
+                throw new IllegalArgumentException("Type of the contact doesn't match");
```
I assume it should be a client error.

kh0ma commented on home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/ContactServiceImpl.java:
```diff
@@ -0,0 +1,104 @@
+        }
+    }
+
+    private ContactDto checkContactType(Contact contact, ContactDto updateContactDto) {
```
`updateContact(`

kh0ma commented on home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/ContactServiceImpl.java:
```diff
@@ -0,0 +1,104 @@
+        return mapper.convert(email, EmailContactDto.class);
+    }
+
+    @Transactional
```
Why do we need a transaction here?

Abhai2016 replied:
Same as above about this transaction. Don't really need it anymore.

kh0ma commented on home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/ContactServiceImpl.java:
```diff
@@ -0,0 +1,104 @@
+    public void deactivateContact(Long id) {
+        contactRepository.findById(id)
+                .orElseThrow(() -> new NotFoundHomeException("Can't find contact with given ID:" + id));
+        contactRepository.deleteById(id);
```
Do we really want completely delete it? As we decided previously we should follow enabled/disabled approach so we will be able to keep historical information.

Abhai2016 replied:
Yeah, sure. Done.

kh0ma commented:
Please fix BUG detected by sonar https://sonarcloud.io/project/issues?id=ita-social-projects-home&open=AXfouTzodl3lZ-ej0CSH&pullRequest=179&resolved=false&types=BUG

Commit:
Issue#134: Made changes which were requested in pull request

Commit:
Issue#134: Fixed some code smells

Commit:
Issue#134: Fixed other code smells

Commit:
Issue#134: Fixed sonar's bugs and code smells from dev

Commit:
Issue#134: Fixed checkstyle errors

Commit:
Issue#134: Fixed sonar bugs and code smells

Commit:
Issue#134: Fixed sonar's bug and code smells

Commit:
Issue#134: More code smells fixed

Commit:
Merge branch 'dev' into Issue#134-Add-Contact-Entity-To-User

Commit:
Issue#134: Added email and password for updating

Commit:
Issue#134: Added contact to updateUser method

Commit:
Issue#134: Deleted updating contacts possibilty in updateUser

VadymSokorenko committed:
Merge branch 'Issue#134-Add-Contact-Entity-To-User' of F:\Java\Workspace\Home with conflicts.

VadymSokorenko committed:
Merge branch 'Issue#134-Add-Contact-Entity-To-User' of https://github.com/ita-social-projects/Home into Issue#134-Add-Contact-Entity-To-User

VadymSokorenko committed:
Issue#134: Added ContactApi tests, refactored UserApi and NewsApi tests.

VadymSokorenko committed:
Issue#134: Resolved conflicts.

VadymSokorenko committed:
Issue#134: code refactoring and minor changes.

- Added check for existing "main" email or phone user contact when adding new contact.
- Renamed CRUD operations in openapi specification.
- User emails are now randomly generated in User API tests.
- Fixed query contacts by ContactType.
- Fixed CheckStyle violations.
- Optimized imports.
- Added Intellij IDEA generated native Code Style. Updated home-docs/code-style.md.
- CheckStyle plugin version updated to 8.41.

VadymSokorenko committed:
Issue#134: Added Cooperation sequence.

VadymSokorenko committed:
Issue#134: Fixed sonar bugs and code smells.

kh0ma commented on .github/home-code-style.xml:
```diff
@@ -0,0 +1,48 @@
+      <option name="CONTINUATION_INDENT_SIZE" value="4" />
+    </indentOptions>
+  </codeStyleSettings>
+</code_scheme>
```
`No newline at the end of the file`

Please add newline

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/api/CommonApi.java:
```diff
@@ -6,28 +6,34 @@
 
-@NoArgsConstructor
-public abstract class CommonApi<R extends BaseDto> {
+@Component
```
Do we really need it here? AFAIK spring cannot instantiate abstract classes as well as plain java. Can we try to remove it?

VadymSokorenko replied:
That is right. Removed annotation.

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/config/rsql/ContactsRSQLEndpointConfig.java:
```diff
@@ -0,0 +1,37 @@
+import org.springframework.stereotype.Component;
+
+@Component
+public class ContactsRSQLEndpointConfig implements RSQLEndpointConfig<Contact, ContactQueryParamEnum> {
```
Let's keep it singular `ContactRSQLEndpointConfig`

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/config/rsql/RSQLConfig.java:
```diff
@@ -23,10 +22,11 @@ public void init() {
             Class<? extends BaseEntity> entityClass = config.getQueryConfig().getEntityClass();
             for (Map.Entry<? extends QueryParamEnum, String> entry : config.getMappings().entrySet()) {
-                RSQLJPASupport.addMapping(entityClass, entry.getKey().getParameter(), entry.getValue());
+                io.github.perplexhub.rsql.RSQLCommonSupport
```
Can we use import instead of qualified import?

VadymSokorenko replied:
Moved to import section.

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/config/rsql/RSQLConfig.java:
```diff
@@ -23,10 +22,11 @@ public void init() {
             }
 
-            RSQLJPASupport.addPropertyWhitelist(
+            io.github.perplexhub.rsql.RSQLCommonSupport.addPropertyWhitelist(
```
Can we use import instead of qualified import?

VadymSokorenko replied:
Moved to import section.

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/config/rsql/UserRSQLEndpointConfig.java:
```diff
@@ -18,12 +18,14 @@
         map.put(UserQueryParamEnum.LAST_NAME, "lastName");
+        map.put(UserQueryParamEnum.CONTACT_PHONE, "contacts.phone");
+        map.put(UserQueryParamEnum.CONTACT_EMAIL, "contacts.email");
+
```
Seems like an extra line. Please remove.

kh0ma commented on home-data-migration/src/main/resources/db/changelog/0.0.1/Issue134.xml:
```diff
@@ -0,0 +1,84 @@
+        </addColumn>
+    </changeSet>
+
+    <changeSet id="Issue134.008" author="VadymSokorenko">
```
I don't think we need a procedure for backfilling here as the cooperation table doesn't exist on any env. Please just add a sequence.

VadymSokorenko replied:
Removed procedure.

kh0ma commented on home-data/src/main/java/com/softserveinc/ita/homeproject/homedata/entity/ContactTypeAttributeConverter.java:
```diff
@@ -0,0 +1,38 @@
+        if (attribute == null) {
+            return null;
+        }
+
```
isn't it an extra line?

kh0ma commented on home-data/src/main/java/com/softserveinc/ita/homeproject/homedata/entity/Cooperation.java:
```diff
@@ -0,0 +1,22 @@
+@Getter
+@Setter
+@Entity
+@Table(name = "cooperation")
```
As we agreed on our decision that we use plural names for our DB table names let's follow it.

kh0ma commented on home-data/src/main/java/com/softserveinc/ita/homeproject/homedata/entity/Permission.java:
```diff
@@ -7,20 +7,17 @@
 @SequenceGenerator(name = "sequence", sequenceName = "permissions_sequence")
 public class Permission extends BaseEntity {
 
+
```
extra line

kh0ma commented on home-open-api/pom.xml:
```diff
@@ -63,13 +63,11 @@
                     <validateOperationTag>true</validateOperationTag>
                     <validateOperationSummary>true</validateOperationSummary>
-                    <validateModelPropertiesExample>true</validateModelPropertiesExample>
+                    <validateModelPropertiesExample>false</validateModelPropertiesExample>
```
Why did we disable it? I think we must keep it enabled.

VadymSokorenko replied:
Changed to `true`.

kh0ma commented on home-open-api/pom.xml:
```diff
@@ -63,13 +63,11 @@
+                    <validateModelPropertiesExample>false</validateModelPropertiesExample>
                     <validateModelNoLocalDef>true</validateModelNoLocalDef>
-                    <validateNaming>true</validateNaming>
+                    <validateNaming>false</validateNaming>
```
Why did we disable it? I think we must keep it enabled.

VadymSokorenko replied:
Changed to "true". Added `path`, `parameter`, `header` and `property` naming checks as well.

kh0ma commented on home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/NewsServiceImpl.java:
```diff
@@ -67,29 +72,32 @@ public NewsDto update(Long id, NewsDto newsDto) {
 
         } else {
-            throw new NotFoundHomeException("Can't find news with given ID:" + id);
+            throw new NotFoundHomeException(NOT_FOUND_NEWS + id);
```
Let's use `String.format` instead of concatenation.

VadymSokorenko replied:
Substituted with `String.format` constant.

VadymSokorenko committed:
Issue#134" Post-review changes.

- Implemented requested post-review changes.
- Added openAPI specification validator checks for UnderscoreCase namings.
- Renamed several parameters and properties in openapi.yaml to match UnderscoreCase.
- Added home-docs/home-api-tests.md (running tests guide).
- Changed table name "cooperation" to "cooperations".

VadymSokorenko committed:
Issue#134: Fixing sonar code smells.
